src/runtime/HalideBuffer.h

```diff
@@ -229,42 +229,35 @@ class Buffer {
             alloc = nullptr;
             set_host_dirty(false);
         }
-        decref_dev();
-    }
-
-    void decref_dev() {
-        int new_count = 0;
         if (dev_ref_count) {
-            new_count = --(dev_ref_count->count);
-        }
-        if (new_count == 0) {
-            if (buf.device) {
-                assert(!(alloc && device_dirty()) &&
-                       "Implicitly freeing a dirty device allocation while a host allocation still lives. "
-                       "Call device_free explicitly if you want to drop dirty device-side data. "
-                       "Call copy_to_host explicitly if you want the data copied to the host allocation "
-                       "before the device allocation is freed.");
-                if (dev_ref_count && dev_ref_count->ownership == BufferDeviceOwnership::WrappedNative) {
-                    buf.device_interface->detach_native(nullptr, &buf);
-                } else if (dev_ref_count && dev_ref_count->ownership == BufferDeviceOwnership::AllocatedDeviceAndHost) {
-                    buf.device_interface->device_and_host_free(nullptr, &buf);
-                } else if (dev_ref_count && dev_ref_count->ownership == BufferDeviceOwnership::Cropped) {
-                    buf.device_interface->device_release_crop(nullptr, &buf);
-                } else if (dev_ref_count == nullptr || dev_ref_count->ownership == BufferDeviceOwnership::Allocated) {
-                    buf.device_interface->device_free(nullptr, &buf);
+            int new_count = --(dev_ref_count->count);
```

Comment:
I would have been tempted to think that the previous code setup was better, but only on the assumption that `decref()` was usually inlined and that `decref_dev()` is usually not-inlined. Sounds like that's probably not happening based on your observations with the profiler. I wonder if adding always-inline and never-inline directives would be worth experimenting with.

Reply:
I think either way, there is one non-inline call, and I think it's useful to consolidate the two, so we can better see what the overhead is in a profiler. And if for some reason neither one inlines, then it's an improvement (from two calls to one).